core/utils/codel.h

```diff
@@ -0,0 +1,331 @@
+#ifndef BESS_UTILS_CODEL_H_
+#define BESS_UTILS_CODEL_H_
+
+#include <math.h>
+#include <sys/time.h>
+
+#include <iostream>
+
+#include "./lock_free_queue.h"
+#include "./queue.h"
+
+namespace bess {
+namespace utils {
+/**
+ * Codel(Controlled Delay Management) is an Queue controller based on this
+ * article http://queue.acm.org/detail.cfm?id=2209336
+ *
+ * It provides an active queue management to help prevent bufferbloat by dropping
+ * queue entries at an increasing rate if the delay in the queue is above the
+ * target queue delay. The equation used to calculate drop intervals is based on TCP
+ * throughput response to drop probability.
+ *
+ * Takes Template argument Q which is a templated child class of Queue and T the type
+ * that is going to be enqueued/dequeued.
+ */
+template <template <typename type> class Q, typename T>
+class Codel : public Queue<T> {
+ public:
+  static constexpr uint64_t kDefaultTarget = 5000;  // default delay target for codel
+  static constexpr uint64_t kDefaultWindow = 100000;  // default window size for codel
+  static const int kDefaultSlots = 4096;  // default number of slots in the codel queue
+
+  // A wrapper class around the pointer to track the delay time and whether
+  // the packet can be dropped after dequeueing.
+  struct Wrapper {
+    T* ptr;              // a pointer to be enqueued
```

Comment:
We decided to not change this as llring requires pointers and that will be the primary queue used by codel.